source/adapters/cuda/event.hpp

```diff
@@ -83,6 +83,21 @@ struct ur_event_handle_t_ {
   static ur_event_handle_t
   makeNative(ur_command_t Type, ur_queue_handle_t Queue, CUstream Stream,
              uint32_t StreamToken = std::numeric_limits<uint32_t>::max()) {
+    if (Queue->has_cached_events()) {
+      auto retEvent = Queue->get_cached_event();
+
+      retEvent->Stream = Stream;
+      retEvent->StreamToken = StreamToken;
+      retEvent->CommandType = Type;
+      retEvent->Queue = Queue;
+      retEvent->Context = Queue->Context;
+      retEvent->RefCount = 1;
```

Comment:
Is `RefCount` always going be either 1 or 0 (1 when first created or 0 when reset/destroyed)? I am asking this in relation to whether it will be worth considering incrementing and decrementing the `RefCount` (atomically, i.e. `std::atomic::fetch_add` with `std::memory_order_relaxed`) instead to safely manage copying the `ur_event` object. I haven't checked the constraints and all uses of this object, just a general observation and suggestion.

Reply:
Good question.
RefCount is initialized to 1 in the constructor and indeed 0 when it gets destroyed.
Setting it to 1 in this particular spot should not violate thread-safety, however.
Since the retrieval of events is now atomic inside the queue and this event just lives inside the scope of this function before it's returned, there is no chance for this particular event to be accessed outside and we can therefore modify it's variables safely without creating an invalid state.